shared-bindings/bitmaptools/__init__.c

```diff
@@ -296,6 +296,62 @@ STATIC mp_obj_t bitmaptools_obj_fill_region(size_t n_args, const mp_obj_t *pos_a
 }
 
 MP_DEFINE_CONST_FUN_OBJ_KW(bitmaptools_fill_region_obj, 0, bitmaptools_obj_fill_region);
+//|
+//| def boundary_fill(
+//|        dest_bitmap: displayio.Bitmap,
+//|        x: int, y: int,
+//|        value: int, background_value: int) -> None:
+//|      """Draws the color value into the destination bitmap enclosed
+//|      area of pixels of the background_value color. Like "Paint Bucket"
+//|      fill tool.
+//|
+//|      :param bitmap dest_bitmap: Destination bitmap that will be written into
+//|      :param int x: x-pixel position of the first pixel to check and fill if needed
+//|      :param int y: y-pixel position of the first pixel to check and fill if needed
+//|      :param int value: Bitmap palette index that will be written into the
+//|             enclosed area in the destination bitmap
+//|      :param int background_value: Bitmap palette index that will filled with the
+//|             value color in the enclosed area in the destination bitmap"""
+//|      ...
+//|
+STATIC mp_obj_t bitmaptools_obj_boundary_fill(size_t n_args, const mp_obj_t *pos_args, mp_map_t *kw_args) {
+    enum {ARG_dest_bitmap, ARG_x, ARG_y, ARG_value, ARG_background_value};
+
+    static const mp_arg_t allowed_args[] = {
+        {MP_QSTR_dest_bitmap, MP_ARG_REQUIRED | MP_ARG_OBJ},
+        {MP_QSTR_x, MP_ARG_REQUIRED | MP_ARG_INT},
+        {MP_QSTR_y, MP_ARG_REQUIRED | MP_ARG_INT},
+        {MP_QSTR_value, MP_ARG_REQUIRED | MP_ARG_INT},
+        {MP_QSTR_background_value, MP_ARG_REQUIRED | MP_ARG_INT},
+    };
+    mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
+    mp_arg_parse_all(n_args, pos_args, kw_args, MP_ARRAY_SIZE(allowed_args), allowed_args, args);
+
+    displayio_bitmap_t *destination = MP_OBJ_TO_PTR(args[ARG_dest_bitmap].u_obj); // the destination bitmap
+
+    uint32_t value, color_depth;
+    value = args[ARG_value].u_int;
+    color_depth = (1 << destination->bits_per_value);
+    if (color_depth <= value) {
+        mp_raise_ValueError(translate("value out of range of target"));
+    }
+
+    uint32_t background_value;
+    background_value = args[ARG_background_value].u_int;
+    if (color_depth <= background_value) {
+        mp_raise_ValueError(translate("background value out of range of target"));
+    }
+
+    int16_t x = args[ARG_x].u_int;
```

Comment:
for x and y (and maybe the colors?) is it worth checking they are valid. I didn't set x to 1000 on a 320x display?

Reply:
Yep I do think its worth adding a check to ensure the location is in bounds in the bitmap. Ill try to work that in.

Thanks for taking a look at this, and all of the suggestions

Reply:
commit `029150a` enforces the initial x,y values passed to the function to ensure they are within the bounds of the bitmap.

There seems to be more than meets they eye going on with the color values. There is actually some bounds checking occurring already, but it seems to be based on powers of 2 specifically and not exactly the `value_count` that was passed when initializing the Bitmap. So it will raise an error if you try to use a color index that is "large enough" over the `value_count` but I'm not sure if it's possible to have it raise the error if you try to use an index value that is over `value_count` but under that next smallest power of 2. I think Bitmap would need to be modified somehow to do that if I understand correctly.

This self contained example illustrates how Bitmap behaves with regards to the  `value_count` passed when creating the Bitmap.
```python
bitmap_test = displayio.Bitmap(2, 2, 5)
bitmap_test[0,0] = 15  # no error raised
bitmap_test[0,0] = 16  # raises ValueError "value requires too many bits"
```

Reply:
Range checks look good. The color values are outside my immediate knowledge.